internal/command/deploy/machines.go

```diff
@@ -207,14 +211,129 @@ func (md *machineDeployment) runReleaseCommand(ctx context.Context) error {
 	return nil
 }
 
+func (md *machineDeployment) resolveProcessGroupChanges() ProcessGroupsDiff {
+
+	output := ProcessGroupsDiff{
+		groupsToRemove:        map[string]int{},
+		groupsNeedingMachines: map[string]*appconfig.ProcessConfig{},
+	}
+
+	groupHasMachine := map[string]bool{}
+
+	for _, leasableMachine := range md.machineSet.GetMachines() {
+		mach := leasableMachine.Machine()
+		machGroup := mach.ProcessGroup()
+		groupMatch := ""
+		for name := range md.processConfigs {
+			if machGroup == name {
+				groupMatch = machGroup
+				break
+			}
+		}
+		if groupMatch == "" {
+			output.groupsToRemove[groupMatch] += 1
+			output.machinesToRemove = append(output.machinesToRemove, leasableMachine)
+		} else {
+			groupHasMachine[groupMatch] = true
+		}
+	}
+
+	for name, val := range md.processConfigs {
+		if hasMach, ok := groupHasMachine[name]; !ok || !hasMach {
+			output.groupsNeedingMachines[name] = val
+		}
+	}
+
+	return output
+}
+
+func (md *machineDeployment) warnAboutProcessGroupChanges(ctx context.Context, diff ProcessGroupsDiff) {
+
+	var (
+		io                 = iostreams.FromContext(ctx)
+		colorize           = io.ColorScheme()
+		willAddMachines    = len(diff.groupsNeedingMachines) != 0
+		willRemoveMachines = diff.machinesToRemove != nil
+	)
+
+	if willAddMachines || willRemoveMachines {
+		fmt.Fprintln(io.ErrOut, "Process groups have changed. This will:")
+	}
+	// TODO(ali): Figure out why the bullets aren't colored
+	if willRemoveMachines {
+		bullet := colorize.Red("*")
+		for grp, numMach := range diff.groupsToRemove {
+			pluralS := lo.Ternary(numMach == 1, "", "s")
+			fmt.Fprintf(io.ErrOut, " %s destroy %d \"%s\" machine%s\n", bullet, numMach, grp, pluralS)
+		}
+	}
+	if willAddMachines {
+		bullet := colorize.Green("*")
+
+		for name := range diff.groupsNeedingMachines {
+			fmt.Fprintf(io.ErrOut, " %s create 1 \"%s\" machine\n", bullet, name)
+		}
+	}
+
+}
+
+func (md *machineDeployment) spawnMachineInGroup(ctx context.Context, groupName string) error {
+	fmt.Fprintf(md.io.Out, "No machines in group '%s', launching one new machine\n", md.colorize.Bold(groupName))
+	machBase := &api.Machine{
+		Region: md.appConfig.PrimaryRegion,
+		Config: &api.MachineConfig{
+			Metadata: map[string]string{
+				api.MachineConfigMetadataKeyFlyProcessGroup: groupName,
+			},
+		},
+	}
+	launchInput := md.resolveUpdatedMachineConfig(machBase, false)
```

Comment:
I know ideally we have to test `spawnMachineGroup` but for now I think it can do with a test case for resolveUpdatedMachineConfig and the same input passed here.

Reply:
I had a whole little test suite for this feature that I wrote in rust, so I just ported it to preflight.

I couldn't figure out a good way to test that part of the code without making it brittle (splitting things into tiny utility functions) or easy to fall out of sync (just copying bits of code), so it just makes more sense to me to do a whole end-to-end test rather than just the config resolver.

Reply:
sgtm 👍